app/scripts/background.js

```diff
@@ -184,10 +185,38 @@ let connectExternal;
 browser.runtime.onConnect.addListener(async (...args) => {
   // Queue up connection attempts here, waiting until after initialization
   await isInitialized;
+  const remotePort = args[0];
+  const { sender, name } = remotePort;
 
-  // This is set in `setupController`, which is called as part of initialization
-  connectRemote(...args);
+  const url = sender?.url;
+  const detectedProcessName = url ? getEnvironmentType(url) : '';
+  const processName = name;
+
+  if (
+    processName === 'metamask-contentscript' &&
+    detectedProcessName === 'background'
+  ) {
+    // This is set in `setupController`, which is called as part of initialization
+    connectRemote(...args);
+    return;
+  }
+
+  if (detectedProcessName !== processName) {
+    throw new Error('UNEXPECTED_CONNECTION_ATTEMPT');
+  }
+
+  const connectionId = generateConnectionId(remotePort, processName);
```

Comment:
Could we use the `detectedProcessName` here instead? My thinking is that the detected process name might be more difficult to spoof, because it's derived from the URL which is guaranteed by the browser to be a true reflection of the sender.